pkg/storegateway/bucket.go

```diff
@@ -1043,6 +950,146 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 	return err
 }
 
+// synchronousSeriesSet returns seriesSet that contains the requested series. It returns a cleanup func. The cleanup func
+// should be invoked always when non-nil; even when the returned error is non-nil. synchronousSeriesSet must be called
+// while holding s.mtx.RLock(). synchronousSeriesSet will call s.mtx.RUnlock regardless of its return values
+func (s *BucketStore) synchronousSeriesSet(
```

Comment:
Instead of returning `*queryStats`, I would take in input `*safeQueryStats` like done in other functions (so we follow the same pattern).